zipline/utils/tradingcalendar.py

```diff
@@ -12,393 +11,250 @@
+
+        self.schedule = DataFrame(
+            index=_all_days,
+            columns=['market_open', 'market_close'],
+            data={
+                'market_open': _opens,
+                'market_close': _closes,
+            },
+            dtype='datetime64[ns]',
+        )
+
+        self.first_trading_day = _all_days[0]
+        self.last_trading_day = _all_days[-1]
+
+    def _special_dates(self, calendars, ad_hoc_dates, start_date, end_date):
+        """
+        Union an iterable of pairs of the form
+
+        (time, calendar)
+
+        and an iterable of pairs of the form
+
+        (time, [dates])
+
+        (This is shared logic for computing special opens and special closes.)
+        """
+        tz = self.native_timezone
+        _dates = DatetimeIndex([], tz='UTC').union_many(
+            [
+                holidays_at_time(calendar, start_date, end_date, time_, tz)
+                for time_, calendar in calendars
+            ] + [
+                days_at_time(datetimes, time_, tz)
+                for time_, datetimes in ad_hoc_dates
+            ]
+        )
+        return _dates[(_dates >= start_date) & (_dates <= end_date)]
+
+    def _special_opens(self, start, end):
+        return self._special_dates(
+            self.special_opens_calendars,
+            self.special_opens_adhoc,
+            start,
+            end,
+        )
+
+    def _special_closes(self, start, end):
+        return self._special_dates(
+            self.special_closes_calendars,
+            self.special_closes_adhoc,
+            start,
+            end,
+        )
+
+    @abstractproperty
+    def native_timezone(self):
+        """
+        Native timezone for the exchange.
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def open_time(self):
+        """
+        datetime.time at which the exchange opens on normal days
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def close_time(self):
+        """
+        Hour, in UTC, at which the exchange opens on no.
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def holidays_calendar(self):
+        """
+        An instance of pd.AbstractHolidayCalendar representing
+        regularly-occurring holidays.
+        """
+        raise NotImplementedError()
+
+    @abstractproperty
+    def special_opens_calendars(self):
+        """
+        Iterable of pairs of the form (datetime.time, AbstractHolidayCalendar).
+
+        Defines dates and times on which the calendar regularly opens at a
+        nonstandard time.
+        """
+        return ()
+
+    @abstractproperty
+    def special_closes_calendars(self):
+        """
+        Iterable of pairs of the form (datetime.time, AbstractHolidayCalendar).
+
+        Defines dates and times on which the calendar regularly closes at a
+        nonstandard time.
+        """
+        return ()
+
+    @abstractproperty
+    def holidays_adhoc(self):
```

Comment:
Should this have another name then `holidays`? This is for cases like Hurricane Sandy, correct? So maybe something like `market_closed_adhoc`?